src/semantics/expression_builder.ts

```diff
@@ -1270,6 +1281,8 @@ export function newCastValue(
                     );
                 }
             }
+            // when the 'meta' of r-value is modified, the 'typeId' of r-value also needs to be modified.
+            from_value.type.typeId = type.typeId;
```

Comment:
You put this line in the for-loop scope, this means you can ensure all object literal to become the same shape with the assigned interface type, but seems this is not always guaranteed. e.g.:
``` TypeScript
interface I {
    x?: number
    y?: boolean
    z?: stirng
}

let obj : I = {
    y: true,
    t: 'hello',
}
```
You can add the missing `x` and `z` field, and you can adjust `y`'s type, but you can't remove the field `t`, in this case the type id is not the same with the interface `I`. So I think you need more strict rule to decide when to change the typeid.